BUG/ENH: Removed non-standard scaling of the covariance matrix and added option to disable scaling completely.

wummo opened the pull request:
Fixes #11196 

As discussed in the bug report polyfit uses a non-standard scaling factor for the covariance matrix, this is corrected. 

Furthermore, an option is added to disable the scaling of the covariance matrix completely. It would be useful in occasions, where the weights are given by 1/sigma with sigma being the (known) standard errors of (Gaussian distributed) data points, in which case the unscaled matrix is already a correct estimate for the covariance matrix.

wummo committed:
Removed non-standard scaling of the covariance matrix and added option to disable scaling completely.

mattip commented:
Thanks for the pull request, and welcome!

Needs a mention in `doc/release/1.15.0-notes.rst` under the `Changes` section, as well as a mention in the `Improvements` section for the additional option. 

Also tests are missing for the new option. It would be nice to have a "demonstration of desired behavior" type of test that simply demonstrates the power of the new option, as well as a test for any new error modes. For instance, what happens if `w` is None but `absolute_weights` is `True`?

Since we are changing default behavior, a heads-up to the numpy-discussion mailing list with a link to this commit is also necessary.

wummo commented:
Hi Matti,

thanks for the welcome. 

I changed the release notes and added  some notes in the sections "changes" and "improvements". 
Also, I added  a check if `absolute_weights` is set to `True` but `w` is empty. In this case a `ValueError` is thrown. Another possible clash of options might be `absolute_weights=True` but `cov=False` but I believe that this can be neglected. Finally, I introduced 2 tests, one to check if the covariance matrix is calculated correctly and another one the see if `ValueError` is thrown. Lastly, I wrote a small note to numpy-discussions about the change in default behavior that comes with this patch. 

Best,
Andreas

mhvk commented on numpy/lib/polynomial.py:
```diff
@@ -423,6 +424,19 @@ def polyfit(x, y, deg, rcond=None, full=False, w=None, cov=False):
     cov : bool, optional
         Return the estimate and the covariance matrix of the estimate
         If full is True, then cov is not returned.
+    absolute_weights: bool, optional
```
I realize it amount to bike-shedding, but I find this name confusing, since I've never encountered this term. If we stick close to this, I'd very much prefer a `relative_weigths=True` since I think that more clearly indicates that there is something weird about the weights.

But really what this does is forcing the reduced chi2 to unity, so maybe that is what the parameter name should reflect? Indeed, in the actual code, the weights are not used at all. Now `force_redchi2_to_unity` is a bit long...  Maybe `rescale_covariance=True`? Or just `cov_scale` or `scale_cov`?

jotasi replied:
Just to explain the current choice of `absolute_weights`: It was suggested by @josef-pkt on the [mailing list](http://numpy-discussion.10968.n7.nabble.com/Inconsistent-results-for-the-covariance-matrix-between-scipy-optimize-curve-fit-and-numpy-polyfit-td45582.html). It is the analogue to `scipy.optimize.curve_fit`'s `absolute_sigma` parameter. Its name was decided on in the lengthy discussion within [this PR](https://github.com/scipy/scipy/pull/448#issuecomment-19677564), which was continued in [this PR](https://github.com/scipy/scipy/pull/3098). At least I somewhat like the analogy to the `curve_fit` terminology but I don't know whether this really is a valid argument here.

mhvk replied:
p.s. In the docstring proper, be careful with single back quotes - with those, there should be an actual link target, i.e., something like `False` works because it links to the python API.

mhvk replied:
What about this suggestion of using `cov` to indicate whether or not scaling should be done?

wummo replied:
To be honest, I like the simple `relative_weight` better. Like @jotasi already mentioned, it behaves like `absolute_sigma` in `curve_fit`. Still, if you insist, I can implement your proposal. In this case it would be nice, if you could point me to another function with a similar parameter, so I can have a look what type of parameter check is performed.

mhvk commented on numpy/lib/polynomial.py:
```diff
@@ -552,6 +566,8 @@ def polyfit(x, y, deg, rcond=None, full=False, w=None, cov=False):
         raise TypeError("expected 1D or 2D array for y")
     if x.shape[0] != y.shape[0]:
         raise TypeError("expected x and y to have same length")
+    if absolute_weights and (w is None):
```
I don't see why this is necessary: the rescaling could be done or omitted (as is arguably meaningful) independent of whether weights are present. I'd remove this.

jotasi replied:
Wouldn't omitting the rescaling without specifying weights effectively mean that all points' standard deviations are considered 1, or did I misunderstand that?

mhvk replied:
Yes, for normal distributions that would be the case. But mostly I see no reason to force a user to pass on `w=np.ones(y.shape[0])` when the flag is set. The default is not really no weights, but weight equals 1.

wummo replied:
In the case of relative weights (scaling), having all weights set to one means that the error on all data points is of equal magnitude. In the new case (absolute weights, no scaling), I'm not sure if this is any sensible default. Why would a data point have the error sigma==1? I guess, that would be just by coincidence or in special cases where you draw from a distribution with known width (like your unit test example)?

mhvk replied:
@wummo - it indeed implies `sigma=1` - which I agree is not necessarily all that meaningful, but I don't see a reason to specifically forbid someone from entering it - it just makes the code longer and more complex for no benefit.

wummo replied:
OK, here is my last argument: Isn't sigma==1 a detail of the implementation that might (maybe) change? Then, giving no weights is something like "undefined behavior".

mhvk replied:
But it isn't ;-) After all, this is a possibly weighted least-squares fit, not a chi2 one, and the meaning is clear without the weights (the meaning of the covariance admittedly less so, but I don't think one has to hand-hold people that much).

wummo replied:
In my opinion, the test is useful, but having very little experience with numpy development/guidlines, I followed your hints and removed the check and the corresponding unit test.

mhvk commented on numpy/lib/polynomial.py:
```diff
@@ -590,14 +606,13 @@ def polyfit(x, y, deg, rcond=None, full=False, w=None, cov=False):
     elif cov:
         Vbase = inv(dot(lhs.T, lhs))
         Vbase /= NX.outer(scale, scale)
-        # Some literature ignores the extra -2.0 factor in the denominator, but
```
Wouldn't it be nice if this comment included a reference... cc @teoliphant who git assigns the blame to. 

In any case, should we add a note that this was different here? Just in case someone later decides that really Bayesian analysis insists there should be a `-2`... (to help ensure this doesn't happen, I also suggest a 0-D test case below - any changes from the now-corrected scaling will make that fail!)

wummo replied:
I added a small note about the removal of the "-2.".

mhvk commented on numpy/lib/polynomial.py:
```diff
@@ -590,14 +606,13 @@ def polyfit(x, y, deg, rcond=None, full=False, w=None, cov=False):
+        else:
+            if len(x) <= order:
+                raise ValueError("the number of data points must exceed order "
+                                 "for estimate the covariance matrix")
```
This error message is not correct any more, it only needs to be true if rescaling is done. Maybe just replace "for estimate" (weird grammar anyway) with "to scale"

mhvk replied:
Still to be done: "for estimate" -> "to scale" (or "in order to be able to scale"

mhvk commented on numpy/lib/tests/test_polynomial.py:
```diff
@@ -136,27 +136,39 @@ def test_polyfit(self):
-                      [0, 1, 3], [0, 1, 3], deg=0, cov=True)
+                      [1], [1], deg=0, cov=True)
+
+        # Check exception when option absolute_weights is True, but  no weights
```
This would need to be removed again...

mhvk commented on numpy/lib/tests/test_polynomial.py:
```diff
@@ -136,27 +136,39 @@ def test_polyfit(self):
+        assert_raises(ValueError, np.polyfit,
+                      x, y, deg=1, absolute_weights=True)
 
         # check 1D case
```
Could we add a "check 0D case"? I ask since for that case we know exactly what the answer should be. (True also for higher-D, but less obvious): just pass in normally distributed data and check that `sigma_mean = sqrt(cov) = sigma/sqrt(n)`. E.g.,
```
np.random.seed(123)
y = np.random.normal(size=(4, 10000))
mean, cov = np.polyfit(np.zeros(y.shape[0], y, deg=0, cov=True)
# Should get sigma_mean = sigma/sqrt(N) = 1./sqrt(4) = 0.5.
assert_allclosel(mean.std(),  0.5, atol=0.01)
assert_allclose(np.sqrt(cov.mean()), 0.5, atol=0.01)
``` 
(Obviously, the last test would fail on current master, which goes to show that the -2 is wrong at least for the simplest case.)

mhvk replied:
Just realized the above is missing a test of the unscaled version:
```
mean, cov_noscaling = np.polyfit(np.zeros(y.shape[0], y, deg=0, cov=True, cov_scale=False)
# without scaling, for 0D: <cov> = (N-dof)/N = 3/4.
assert_allclose(cov.mean(), 0.75, atol=0.01)
```

wummo replied:
Sorry, but I'm not sure if I understand the tests correctly. My calculations go as follows:
1) absolute_weights = False:
var(a) = var(y) / n_dof
(biased) variance of drawing 4 Gaussian random numbers with width 1: var_4 = (n-1)/n*var = 3/4
therefore: var(a)= var_4 / (n-1) = var / n 
or: sigma(a) = sqrt(sigma/n) = 1/sqrt(4) = 1/2

2) absolute_weights = True:
ar(a) = 1/S, with S=sum_{i=1}^n (1/var_i) 
with var_1 = var_2 = ... var_n = var: S = n / var
or: var(a) = var / n
then: sigma(a) = sqrt(var/ n)  
with var=1: sigma_1(a) = sqrt(1/4) = 1/2
with var=3/4: sigma_3/4(a) = sqrt(3/4/4) = sqrt(3)/4 = 0.43301270...

For the relative weights (1) we seem to agree while for the absolute weights (2), there is a difference:

```python
import numpy as np

np.random.seed(123)
y = np.random.normal(size=(4, 10000))
x = np.zeros(y.shape[0])

print("*** relative weights ***")
mean, cov = np.polyfit(x, y, deg=0, cov=True, absolute_weights=False)
print("expected: 1/sqrt(4)=0.5")
print("np.sqrt(cov.mean())={}".format(np.sqrt(cov.mean())))
print("")
print("")
print("*** absolute weights***")
print("a) var_i=1")
var = np.ones(y.shape[0])
mean, cov = np.polyfit(
    x, y, deg=0, cov=True, w=1 / np.sqrt(var), absolute_weights=True)
print("expected: 1/sqrt(4)=0.5")
print("np.sqrt(cov.mean())={}".format(np.sqrt(cov.mean())))
print("")

print("b) var_i=3/4")
var = np.full(y.shape[0], 3 / 4)
mean, cov = np.polyfit(
    x, y, deg=0, cov=True, w=1 / np.sqrt(var), absolute_weights=True)
print("expected: sqrt(3)/4={}".format(np.sqrt(3) / 4))
print("np.sqrt(cov.mean())={}".format(np.sqrt(cov.mean())))
```
gives
```
*** relative weights ***
expected: 1/sqrt(4)=0.5
np.sqrt(cov.mean())=0.5008913441736218


*** absolute weights***
a) var_i=1
expected: 1/sqrt(4)=0.5
np.sqrt(cov.mean())=0.5

b) var_i=3/4
expected: sqrt(3)/4=0.4330127018922193
np.sqrt(cov.mean())=0.43301270189221924
```

mhvk replied:
Yes, I'm an idiot! For the unscaled case, of course, the `sqrt(cov.mean())` again should be 0.5, since the weights are 1 and by construction the reduced chi2 is 1 too.

So, I think the tests might be better as:
```
np.random.seed(123)
y = np.random.normal(size=(4, 10000))
mean, cov = np.polyfit(np.zeros(y.shape[0], y, deg=0, cov=True)
# Should get sigma_mean = sigma/sqrt(N) = 1./sqrt(4) = 0.5.
assert_allclosel(mean.std(),  0.5, atol=0.01)
assert_allclose(np.sqrt(cov.mean()), 0.5, atol=0.01)
# Without scaling, since reduced chi2 is 1, the result should be the same.
mean, cov = np.polyfit(np.zeros(y.shape[0], y, deg=0, cov=True, absolute_weights=True)
assert_allclosel(mean.std(),  0.5, atol=0.01)
assert_allclose(np.sqrt(cov.mean()), 0.5, atol=0.01)
# If we estimate our errors wrong, no change with scaling:
w = np.full(y.shape[0], 1./0.5**2)
mean, cov = np.polyfit(np.zeros(y.shape[0], y, deg=0, cov=True)
assert_allclosel(mean.std(),  0.5, atol=0.01)
assert_allclose(np.sqrt(cov.mean()), 0.5, atol=0.01)
# But if we do not scale, our estimate for the error in the mean will differ.
mean, cov = np.polyfit(np.zeros(y.shape[0], y, deg=0, cov=True, absolute_weights=True)
assert_allclosel(mean.std(),  0.5, atol=0.01)
assert_allclose(np.sqrt(cov.mean()), 0.25, atol=0.01)
```

wummo replied:
I guess, the 3rd test should look like this?

```Python
# If we estimate our errors wrong, no change with scaling:
w = np.full(y.shape[0], 1./0.5**2)
mean, cov = np.polyfit(np.zeros(y.shape[0]), y, w, deg=0, cov=True)
assert_allclosel(mean.std(),  0.5, atol=0.01)
assert_allclose(np.sqrt(cov.mean()), 0.5, atol=0.01)
```
(the weights `w` were passed to np.polyfit)

mhvk replied:
@wummo - yes, that was the idea, sorry for writing tests on the fly, without actually checking them!

mhvk reviewed (changes requested):
Overall, definitely a good idea, but I think the name should reflect what is actually done more closely.

Also, definitely include the test!

eric-wieser commented on numpy/lib/polynomial.py:
```diff
@@ -590,14 +606,13 @@ def polyfit(x, y, deg, rcond=None, full=False, w=None, cov=False):
-                             "for Bayesian estimate the covariance matrix")
-        fac = resids / (len(x) - order - 2.0)
+        if absolute_weights:
+            fac = 1.
```
Best to just use `1` here - it cooperates better with `Decimal`, if that ever becomes supported

wummo replied:
OK, I changed this.

mhvk commented on numpy/lib/polynomial.py:
```diff
@@ -590,14 +606,16 @@ def polyfit(x, y, deg, rcond=None, full=False, w=None, cov=False):
+                                 "for estimate the covariance matrix")
+            # note, this used to be: fac = resids / (len(x) - order - 2.0)
+            # it was deciced that the "- 2" (originally justified by "Bayesian
+            # uncertainty analysis") is not was the user expects
```
Add "(see gh-<number-of-issue> and gh-11197)"

mhvk replied:
For some reason my comment was weird here: should have been "gh-11196 and gh-11197"

mhvk commented:
Looks good, except for the two remaining issues:
1. Do we want to forbid getting an unscaled covariance without weights? I feel this is unnecessary hand-holding.
2. Do we want to fold in the `absolute_weights` into the `cov` boolean, by allowing a string value? Personally, I find routines whose outputs depend on flags super-annoying to start with, and adding even more flags seems silly, so I'd prefer just a single flag that tells what type of covariance to return (where `False` is "don't bother, I don't need them).

mhvk commented:
@wummo - this function is a bit of a mess already... And IIRC it is in fact recommended to use `np.polynomial.polynomial.polyfit`, since it presents coefficients in more logical order, but which does not even have `cov`. Adding another argument makes it deviate even more, but I'm certainly not stuck on sticking to one argument. Let me ping @charris, since he has much more experience with the `polynomial` stuff.

josef-pkt commented:
One argument in favor of rolling it into a trivariate `cov` is that users only have to change one flag to get fixed scale.

(I find it annoying in some cases in statsmodels, where we have one flag to switch away from the default and another flag to choose an option for the alternative. It's easy to forget to switch the first keyword, and I often have to correct my initial code to fix it.)

wummo commented:
@mhvk did you have a chance to contact @charris ?

wummo commented:
In the light of PR #11733, is there still interest to fix this issue here or will this function be dropped eventually anyway, @mhvk @charris?

jsdodge commented:
I just became aware of this issue in `numpy.polyfit` and would strongly recommend including this fix until `numpy.polynomial.polynomial.polyfit` includes an option to return the covariance matrix. Otherwise, users will be stuck with a default that does not provide an easy way to determine parameter uncertainties.

Also, since @josef-pkt expressed confusion about why a user would want this feature in [mailing list discussion](http://numpy-discussion.10968.n7.nabble.com/Inconsistent-results-for-the-covariance-matrix-between-scipy-optimize-curve-fit-and-numpy-polyfit-td45582.html), it's worth noting that it is common practice in physics (my field) to determine the measurement sample variance independently, then treat it as known when fitting a model to data from the same apparatus. [Introductory textbooks](https://global.oup.com/academic/product/measurements-and-their-uncertainties-9780199566327) typically focus on this case.

mattip commented:
Maybe this would be more palatable as two PRs - one to remove the non-standard `-2` and another to add the kwarg. Or is it ready to go in as-is? The `-2` has been a long standing issue (5 years).

mhvk commented:
Sorry that this has slipped so far. I'd still like the opinion of @charris, because it would be good to move this to the polynomial classes. Absent that, I'm happy to merge the `-2` removal, but would prefer the option not to add a new argument, but rather to use a string for the type of covariance one wants.

mhvk commented:
@wummo - fair enough. Yes, please do the string interface and we'll merge this.

mhvk commented on doc/release/1.16.0-notes.rst:
```diff
@@ -239,6 +239,15 @@ single elementary function for four related but different signatures,
 
+New option ``absolute_weights'' in ``np.polyfit''
+-------------------------------------------------
+Like ``absolute_sigma'' in ``scipy.optimize.curve_fit`` a boolean option
```
Need to change the release notes as well...

mhvk commented on doc/release/1.16.0-notes.rst:
```diff
@@ -239,6 +239,15 @@ single elementary function for four related but different signatures,
+(Gaussian distributed) data points, in which case the unscaled matrix is already
+a correct estimate for the covariance matrix. In case ``absolute_weights'' is set
+to true, but no weights are given, a ``ValueError'' is thrown.
 Detailed docstrings for scalar numeric types
```
Note that the rebase has removed the empty line that should be here.

mhvk commented on doc/release/1.16.0-notes.rst:
```diff
@@ -350,6 +359,13 @@ if ``np.positive(array)`` raises a ``TypeError``. For ``ndarray``
+covariance matrix. Namely, rather than using the standard chisq/(M-N), it
+scales it with chisq/(M-N-2) where M is the number of data points and N is the
+number of parameters.  This scaling is inconsistent with other fitting programs
+such as e.g. ``scipy.optimize.curve_fit`` and was changed to chisq/(M-N).
```
And another empty line to be added back in.

mhvk commented on numpy/lib/polynomial.py:
```diff
@@ -463,9 +463,14 @@ def polyfit(x, y, deg, rcond=None, full=False, w=None, cov=False):
+        except in a relative sense and everything is scaled such that the
+        reduced chi2 is unity. This scaling is omitted if ``cov='unscaled'``,
+        as is relevant for the case that the weights are 1/sigma**2, with
+        sigma known to be a reliable estimate of the uncertainty.
```
Very clear, thanks!

mhvk reviewed (commented):
@wummo - thanks for making the changes. Only some small left-overs...

mhvk commented:
p.s. While making the last changes, could you also rebase & squash the commits? Thanks again, and apologies that this has all taken so long.

charris commented:
I've definitely considered adding a covariance computation to the polynomial package fitting functions, I've done it for myself in practice. I agree that `-2` is a bit too clever, in a case like this it is best to follow conventions that folks are used to. Note that for large data sets it won't make much difference, and for small data sets the estimated covariance will have a large error regardless unless the variance of the measurement errors is known by other means.

For `std` and `var`, the "ddof" parameter the default is 0, which is also non-standard, but unlikely to change. There are reasons for that convention as @rkern has explained, but it is surprising to many.

wummo commented:
@mhvk I fixed the problems with the documentation. If everything looks OK, I will do the rebase & squash.

mhvk commented on doc/release/1.16.0-notes.rst:
```diff
@@ -239,6 +239,17 @@ single elementary function for four related but different signatures,
+-----------------------------------------------------------
+
+A further possible value has been added to the ``cov'' parameter of the
+``np.polyfit`` function. With ``cov=unscaled`` the scaling of the covariance
```
One last small thing: missing quotes around unscaled, i.e., ``cov='unscaled'``

mhvk commented:
Looks good modulo the missing quotes. Please go ahead and rebase/squash as well, and I'll merge.

mhvk commented:
@wummo - I hadn't seen that the branch was pushed - now all is OK so I'll merge. Thanks for the contribution and more thanks for your patience!

wummo commented:
Its great that it got merge. Thanks a lot for your help.
